src/common/components/select-repository-list/index.js

```diff
@@ -50,99 +50,104 @@ export class SelectRepositoryListComponent extends Component {
     </span>;
   }
 
+  // TODO this can be moved too select-repository-row as the error prop is on
+  // the repository object passed to it
   getErrorMessage(error) {
-    if (error) {
-      const payload = error.json.payload;
+
+    const { payload = undefined } = error;
+
+    if (payload) {
       if (payload.code === SNAP_NAME_NOT_REGISTERED_ERROR_CODE) {
         return this.getSnapNotRegisteredMessage(payload.snap_name);
       } else {
-        return error.message;
+        return payload.message;
       }
     }
   }
 
-  renderRepository(repo) {
-    const { fullName, enabled } = repo;
-    const status = this.props.repositoriesStatus[fullName] || {};
-    const { selectedRepos } = this.props.selectRepositoriesForm;
+  renderRepository(id) {
+    const repository = this.props.entities.repos[id];
+    const { full_name } = repository;
+
+    const isEnabled = !!this.props.enabledRepositories[id];
 
     return (
       <SelectRepositoryRow
-        key={ `repo_${repo.fullName}` }
-        repository={ repo }
-        buttonLabel={ status.isFetching ? 'Creating...' : 'Create' }
-        buttonDisabled={ status.isFetching }
-        onChange={ this.onSelectRepository.bind(this, repo) }
-        errorMsg= { this.getErrorMessage(status.error) }
-        checked={ hasRepository(selectedRepos, repo) }
-        isEnabled={ enabled }
+        key={ `repo_${full_name}` }
+        repository={ repository }
+        onChange={ this.onSelectRepository.bind(this, id) }
+        errorMsg={ repository.__error && this.getErrorMessage(repository.__error) }
+        isEnabled={ isEnabled }
       />
     );
   }
 
-  onSelectRepository(repository) {
-    this.props.dispatch(toggleRepository(repository));
+  onSelectRepository(id) {
+    this.props.dispatch(toggleRepositorySelection(id));
   }
 
+  // XXX it's not a submit, there's no form, rename to ~handleAddButtonClick
   onSubmit() {
-    const { selectedRepos } = this.props.selectRepositoriesForm;
-    if (selectedRepos.length) {
-      this.props.dispatch(createSnaps(selectedRepos));
+    const { repositoriesToBuild } = this.props;
+
+    // TODO else "You have not selected any repositories"
+    if (repositoriesToBuild.length) {
+      this.props.dispatch(buildRepositories(repositoriesToBuild));
     }
   }
 
   onPageLinkClick(pageNumber) {
     this.props.dispatch(fetchUserRepositories(pageNumber));
   }
 
-  filterEnabledRepos(repositories) {
-    const { success, snaps } = this.props.snaps;
-
-    if (success && snaps.length) {
-      for (let i = repositories.length; i--;) {
-        for (let j = snaps.length; j--;) {
-          const enabledRepo = snaps[j].git_repository_url;
-
-          if (enabledRepo === repositories[i].url) {
-            repositories[i].enabled = true;
-            break;
-          } else {
-            repositories[i].enabled = false;
-          }
-        }
-      }
+  pageSlice(array, pageLinks) {
```

Comment:
@earnubs So if I understand it correctly, right now, with our current "First, Prev, Next, Last" pagination ids may get out of order by using union?

For example user loads first page of repos first (obviously), then uses "Last" to load last page (which is merged by `union` into the result) skipping any pages in between, so moving prev/next to any other pages later means other pages will be merged to the end of ids list while they should be in the middle?